Redirect lxc tx packets directly to the host device

nathanjsweet opened the pull request:
Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

<!-- Description of change -->
 Redirect packets that are egressing from a container directly to the host network device so that they can skip most of the network stack.

```release-note
Added fib lookups to egressing container traffic to route it to the correct device more quickly than traversing the Linux network datapath would by default.
```

brb commented on bpf/bpf_netdev.c:
```diff
@@ -324,6 +349,13 @@ static __always_inline int handle_ipv4(struct __ctx_buff *ctx,
+        /* Skip the stack, if we can lookup the right interface */
+        int ifindex = fib_lookup_ipv4(ip4);
+        if (ifindex > -1)
+                return redirect(ifindex, 0);
```
Wouldn't this redirect to `cilium_host` instead of a container iface? I think we could rely on the endpoint map for redirection (see `lookup_ip4_endpoint()`) instead of `fib_lookup()`.

nathanjsweet replied:
This is egressing out of the container interface, no?

brb replied:
This is in the `from-netdev` section, which is attached to the tc ingress. Outbound packets from the lxc should hit `to-netdev` (tc egress).

brb commented on bpf/bpf_netdev.c:
```diff
@@ -324,6 +349,13 @@ static __always_inline int handle_ipv4(struct __ctx_buff *ctx,
 			return ret;
 	}
+
+#if defined(NO_REDIRECT)
```
We should be able to get rid of the `NO_REDIRECT` once the lxc outbound path redirects instead of returning `TC_ACT_OK`.

nathanjsweet replied:
I think the worry is that the fib lookup may fail still.

nathanjsweet committed:
Directly redirect packets that are egressing from lxc to
the appropriate network device, based on a fib lookup in
the neighbor's table. This should provide some performance
improvements by skipping most of the network stack between
devices.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

qmonnet commented on bpf/bpf_lxc.c:
```diff
@@ -365,6 +365,10 @@ static __always_inline int ipv6_l3_from_lxc(struct __ctx_buff *ctx,
+        int ifindex = fib_lookup_ipv6(ip6);
+        if(ifindex > -1)
+                return redirect(ifindex, 0);
+
```
I don't think we have an explicit coding style, but let's try to keep formatting consistent:
* Please move variable declaration (`int ifindex`) to the top of the function.
* Leave a space between `if` and the opening parenthesis.
* Please fix your indents. They should be tabs (equivalent to 8 spaces).

qmonnet commented on bpf/lib/ipv4.h:
```diff
@@ -134,4 +134,22 @@ ipv4_handle_fragment(struct __ctx_buff *ctx,
+
+	err = fib_lookup(ctx, &fib_params, sizeof(fib_params),
+		    BPF_FIB_LOOKUP_DIRECT | BPF_FIB_LOOKUP_OUTPUT);
+        if (err == 0)
```
`if (!err)`

qmonnet reviewed (changes requested):
A few suggestions on coding style.

Please also add a commit object, and fix your commit log (it does not describe the actual commit).

brb reviewed (commented):
Have you checked whether a neigh entry for a default route always exist? If not, then we should probably do a periodical arping.
